lib.network: add ipv6 parser

woojiq opened the pull request:
### Description of changes

* Add a library function to parse and validate an IPv6 address from a
string. It can parse the first two versions of an IPv6 address according
to https://datatracker.ietf.org/doc/html/rfc4291#section-2.2. The third
form "x:x:x:x:x:x.d.d.d.d" is not yet implemented.
* Add shell script network.sh to test IPv6 parser functionality.
* Copy fromHexString function from the https://github.com/NixOS/nixpkgs/pull/266705

Mentor: @Janik-Haag 

### Things done
- [x] Tested, as applicable:
* [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
* and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
* or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).
- [x] Formatted using nixfmt-rfc-style

Add a 👍 [reaction](https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to [pull requests you find important](https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

lucasew committed:
lib: add fromHexString

Co-authored-by: lucasew <lucas59356@gmail.com>

woojiq committed:
gitignore: add .helix config directory

mweinelt commented on lib/network/default.nix:
```diff
@@ -0,0 +1,107 @@
+{
+  ipv6 = {
+    /**
+      Normalize an IPv6 address by removing the "::" compressions and padding
+      them with the necessary number of zeros to create the correct address. Converts
+      an address from the string to the list of strings which then can be parsed using
+      `_parseNormalized`. Throws an error when the address is malformed.
```
Please talk about expanding and compressing, as that is the official lingo used in the RFC 4291. And make sure to reference RFC as the canonical source for these representations.

> Expand an IPv6 Address according to RFC 4291, by doing this and that

Janik-Haag commented on lib/network/default.nix:
```diff
@@ -0,0 +1,107 @@
+
+      # Type: String -> [ String ]
+    */
+    _normalizeIpv6 =
```
```suggestion
    _normalizeIp =
```
This is already defined in the ipv6 scope, so I think it's preferable to have:
```nix
lib.network.ipv6._normalizeIp
```
over
```nix
lib.network.ipv6._normalizeIpv6
```

Janik-Haag commented on lib/network/default.nix:
```diff
@@ -0,0 +1,107 @@
+{ lib }:
+{
+  ipv6 = {
```
for ergonomics it would be nicer to just use v6 and v4 but I don't have hard feelings here.

woojiq replied:
Personally, I think the "ip" prefix gives more clarity.
`network.v6.func` looks weird.

mweinelt replied:
I agree, proper lingo everywhere is what we should aim for.

Janik-Haag commented on lib/network/default.nix:
```diff
@@ -0,0 +1,107 @@
+          piecesNoEmpty = lib.lists.remove "" pieces;
+          piecesNoEmptyLen = builtins.length piecesNoEmpty;
+          emptyCnt = builtins.length pieces - builtins.length piecesNoEmpty;
+          hasPrefix = lib.strings.hasPrefix doubleColon addr;
+          hasSuffix = lib.strings.hasSuffix doubleColon addr;
+          hasInfix = lib.strings.hasInfix doubleColon addr;
```
I think we should keep it simple and just write "::" over introducing a extra binding, because it makes the code easier to read since you don't have to look up how doubleColon.

```suggestion
          hasPrefix = lib.strings.hasPrefix "::" addr;
          hasSuffix = lib.strings.hasSuffix "::"  addr;
          hasInfix = lib.strings.hasInfix "::" addr;
```

Janik-Haag commented on lib/network/default.nix:
```diff
@@ -0,0 +1,107 @@
+          pieces = lib.strings.splitString ":" addr;
+          piecesNoEmpty = lib.lists.remove "" pieces;
+          piecesNoEmptyLen = builtins.length piecesNoEmpty;
+          emptyCnt = builtins.length pieces - builtins.length piecesNoEmpty;
```
emptyCnt is only used once in line 38, so I think we should move it directly into the local scope. I also think that we should call this emptyCount since everything else is written out and it's only internal anyways.

Janik-Haag commented on lib/network/default.nix:
```diff
@@ -0,0 +1,107 @@
+      assert lib.asserts.assertMsg (
+        builtins.length addr == 8
+      ) "_parseNormalized expected to recieve normalized IPv6 address";
+      builtins.map (el: u16FromHexStr el) addr;
```
```suggestion
      builtins.map (piece: u16FromHexStr piece) addr;
```

We should keep the same language everywhere to avoid confusion.

Janik-Haag commented:
To address your questions:

> Do I need to somehow hide the internal functions more than just underscoring them?

Yes currently they would all get exposed, you have to put them into a let binding to make them truly internal. Take a look at https://github.com/NixOS/nixpkgs/blob/master/lib/fileset/default.nix#L103 as an example.

> Are the doc comments in the correct format (already existing files have different style)?

Some of them are some of them aren't, please take a look at the [example](https://github.com/nix-community/nixdoc?tab=readme-ov-file#example)

> Do I need to pull the ipv4 parser implementation to merge them together? https://github.com/NixOS/nixpkgs/pull/299409

no, we should keep them separate and preferably go for small prs, e.g. once the parser is fine merge this pr and then do another v6 one for cidr related functions.

woojiq commented on lib/tests/network.sh:
```diff
@@ -0,0 +1,84 @@
+
+piecesMax=8
+
+expectSuccess 'ipv6.fromString "0:0:0:0:0:0:0:0"'                         '[0,0,0,0,0,0,0,0]'
+expectSuccess 'ipv6.fromString "000a:000b:000c:000d:000e:000f:ffff:aaaa"' '[10,11,12,13,14,15,65535,43690]'
+expectSuccess 'ipv6.fromString "::"'                                      '[0,0,0,0,0,0,0,0]'
+expectSuccess 'ipv6.fromString "::0000"'                                  '[0,0,0,0,0,0,0,0]'
+expectSuccess 'ipv6.fromString "::1"'                                     '[0,0,0,0,0,0,0,1]'
+expectSuccess 'ipv6.fromString "::ffff"'                                  '[0,0,0,0,0,0,0,65535]'
+expectSuccess 'ipv6.fromString "::000f"'                                  '[0,0,0,0,0,0,0,15]'
+expectSuccess 'ipv6.fromString "::1:1:1:1:1:1:1"'                         '[0,1,1,1,1,1,1,1]'
+expectSuccess 'ipv6.fromString "1::"'                                     '[1,0,0,0,0,0,0,0]'
+expectSuccess 'ipv6.fromString "1:1:1:1:1:1:1::"'                         '[1,1,1,1,1,1,1,0]'
+expectSuccess 'ipv6.fromString "1:1:1:1::1:1:1"'                          '[1,1,1,1,0,1,1,1]'
+expectSuccess 'ipv6.fromString "1::1"'                                    '[1,0,0,0,0,0,0,1]'
```
I think we can use reserved ip in examples but for testing purpose reserved ip might not be enough.

woojiq commented:
@Janik-Haag @mweinelt I refactored the code and moved internal functions to internal.nix. Can you review again please?

woojiq committed:
lib.network: ipv6 parser from string

Add a library function to parse and validate an IPv6 address from a
string. It can parse the first two versions of an IPv6 address according
to https://datatracker.ietf.org/doc/html/rfc4291#section-2.2. The third
form "x:x:x:x:x:x.d.d.d.d" is not yet implemented. Optionally parser can accept prefix length (128 is default).

Add shell script network.sh to test IPv6 parser functionality.

Janik-Haag reviewed (changes requested):
I think this is getting into shape. There are some smaller points left to do, and you should add a lot more comments explaining different parts, I think. This part of lib currently is very undocumented internally compared to others like lib.fileset or lib.generators.

woojiq commented on lib/network/internal.nix:
```diff
@@ -0,0 +1,143 @@
+in
+let
+  /**
+    Expand an IPv6 address by removing the "::" compression and padding them
```
RFC 4291:

> In order to make writing addresses containing zero bits easier, a special syntax is available to compress the zeros.

> or in compressed form:

woojiq commented on lib/network/internal.nix:
```diff
@@ -0,0 +1,143 @@
+    */
+    split =
+      apiName: addr:
+      let
+        splitted = strings.splitString "/" addr;
+        splittedLength = length splitted;
+      in
+      if splittedLength == 1 then # [ ip ]
+        {
+          address = parseIpv6FromString apiName addr;
+          prefixLength = 64; # /64 is a standard size IPv6 subnet as defined by the IETF.
+        }
+      else if splittedLength == 2 then # [ ip subnet ]
```
I agree, 128 sounds more logical if prefix length is omitted.

> I think we should be stricter with the user input and don't silently just set some prefix but instead fail saying that it's not a cidr address.

I added one `fromString` function and user can also bot default and cidr notation. I don't see a reason to create two separate API functions like `fromString` and `fromCidrString`, because this can be easily observed from the passed string - if it contains a `/`, then we try to parse it as cidr notation.

woojiq replied:
Ok, when I started working on ipv6 types, I must have understood why we need separate functions for default and cidr notation. In progress.

Janik-Haag commented on lib/network/default.nix:
```diff
@@ -0,0 +1,49 @@
+    /**
+      Creates an `IPv6Address` object from an IPv6 address as a string. If
+      the prefix length is omitted, it defaults to 64. The parser is limited
+      to the first two versions of IPv6 addresses addressed in RFC 4291.
+      The form "x:x:x:x:x:x:d.d.d.d" is not yet implemented. Addresses are
```
>       The form "x:x:x:x:x:x:d.d.d.d" is not yet implemented.

We probably want to be more explicit and say The form... won't be implemented.

Janik-Haag reviewed (commented):
@infinisil as a lib code owner, do you want to give this a review before we merge it?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/showcase-nixpkgs-network-library/48897/1

drupol commented on .gitignore:
```diff
@@ -7,6 +7,7 @@
 .idea/
 .nixos-test-history
 .vscode/
+.helix/
```
I would add that this shouldn't be part of this PR and added into the user global's `.gitignore` file.

Janik-Haag replied:
I think it's reasonable to have this, esp. since we also have it for other editors like .idea, .vscode, etc... but yes I agree it would have been more transparent if we moved it into a separate pr.
